src/Middleware/Rewrite/src/Internal/RedirectToWwwRule.cs

```diff
@@ -12,11 +13,22 @@ namespace Microsoft.AspNetCore.Rewrite.Internal
     public class RedirectToWwwRule : IRule
     {
         public readonly int _statusCode;
+        public string[] _domains = new string[0];
```

Comment:
You don't need to set this to a default value if you take my suggestion on throwing if the domains is an empty array.

Reply:
I initialized the domains array with a default value in case the original constructor is called where we can't throw.

If that's a "deal breaker" we could try a different approach by adding another boolean that defines if the rule should apply to all domains. We can set that boolean to true on the original constructor and use it to skip the check on the domains array when applying the rule: 

```
public class RedirectToWwwRule : IRule
{
    public readonly int _statusCode;
    public readonly string[] _domains;
    public readonly bool _applyToAllDomains;

    public RedirectToWwwRule(int statusCode)
    {
        _statusCode = statusCode;
        _applyToAllDomains = true;
    }
        
    public RedirectToWwwRule(int statusCode, params string[] domains)
    {
        if (domains == null || domains.Length == 0)
        {
            throw new ArgumentNullException(nameof(domains));
        }
        if (domains.Length < 1)
        {
            throw new ArgumentException(nameof(domains));
        }
        _domains = domains;
        _applyToAllDomains = false;
        _statusCode = statusCode;
    }

    public virtual void ApplyRule(RewriteContext context)
    {
        var req = context.HttpContext.Request;
        if (req.Host.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

        if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

        if (!_applyToAllDomains)
        {
            var domainRequiresRedirection = false;
            foreach (var domain in _domains)
            {
                if (domain.Equals(req.Host.Host, StringComparison.OrdinalIgnoreCase))
                {
                    domainRequiresRedirection = true;
                    break;
                }
            }
            if (!domainRequiresRedirection)
            {
                context.Result = RuleResult.ContinueRules;
                return;
            }
        }

        var wwwHost = new HostString($"www.{req.Host.Value}");
        var newUrl = UriHelper.BuildAbsolute(req.Scheme, wwwHost, req.PathBase, req.Path, req.QueryString);
        var response = context.HttpContext.Response;
        response.StatusCode = _statusCode;
        response.Headers[HeaderNames.Location] = newUrl;
        context.Result = RuleResult.EndResponse;
        context.Logger?.RedirectedToWww();
    }
}
```

Reply:
Do that except with a null check instead:

```c#
public class RedirectToWwwRule : IRule
{
    public readonly int _statusCode;
    public readonly string[] _domains;

    public RedirectToWwwRule(int statusCode)
    {
        _statusCode = statusCode;
    }

    public RedirectToWwwRule(int statusCode, params string[] domains)
    {
        if (domains == null)
        {
            throw new ArgumentNullException(nameof(domains));
        }
        _domains = domains;
        _statusCode = statusCode;
    }

    public virtual void ApplyRule(RewriteContext context)
    {
        var req = context.HttpContext.Request;
        if (req.Host.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

        if (req.Host.Value.StartsWith("www.", StringComparison.OrdinalIgnoreCase))
        {
            context.Result = RuleResult.ContinueRules;
            return;
        }

//      if (_domains != null)
//      {
            var requestRequiresRedirection = false;
            foreach (var domain in _domains)
            {
                if (domain.Equals(req.Host.Host, StringComparison.OrdinalIgnoreCase))
                {
                    requestRequiresRedirection = true;
                    break;
                }
            }
            if (!requestRequiresRedirection)
            {
                context.Result = RuleResult.ContinueRules;
                return;
            }
//      }

        var wwwHost = new HostString($"www.{req.Host.Value}");
        var newUrl = UriHelper.BuildAbsolute(req.Scheme, wwwHost, req.PathBase, req.Path, req.QueryString);
        var response = context.HttpContext.Response;
        response.StatusCode = _statusCode;
        response.Headers[HeaderNames.Location] = newUrl;
        context.Result = RuleResult.EndResponse;
        context.Logger?.RedirectedToWww();
    }
}
```

Reply:
Yes! How didn't I see that option... Thanks, I love this solution :)